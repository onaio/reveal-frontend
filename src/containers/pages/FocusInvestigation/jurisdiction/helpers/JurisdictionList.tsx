import { ObjectList } from '@onaio/cbv';
import { Registry } from '@onaio/redux-reducer-registry';
import { FIReasons } from '../../../../../configs/settings';
import { SORT_BY_EFFECTIVE_PERIOD_START_FIELD } from '../../../../../constants';
import {
  InterventionType,
  makePlansArraySelector,
  PlanFilters,
  PlanStatus,
} from '../../../../../store/ducks/plans';

/** Custom implementation of ObjectList where we override the getMapStateToProps method */
export default class JurisdictionList<
  ObjectType,
  ActionType,
  SelectorType,
  PropsType
> extends ObjectList<ObjectType, ActionType, SelectorType, PropsType> {
  /** Custom getMapStateToProps
   * Gets plans for the current jurisdiction
   */
  public getMapStateToProps() {
    return (state: Registry, ownProps: any) => {
      const getPlansArray = makePlansArraySelector(undefined, SORT_BY_EFFECTIVE_PERIOD_START_FIELD);

      const jurisdictionId = ownProps.match.params.jurisdictionId;

      const reactiveFilters: PlanFilters = {
        interventionType: [InterventionType.FI, InterventionType.DynamicFI],
        jurisdictionIds: [jurisdictionId],
        reason: FIReasons[1],
      };

      const routineFilters: PlanFilters = {
        ...reactiveFilters,
        reason: FIReasons[0],
      };

      return {
        completeReactivePlans: getPlansArray(state, {
          ...reactiveFilters,
          statusList: [PlanStatus.COMPLETE],
        }),
        completeRoutinePlans: getPlansArray(state, {
          ...routineFilters,
          statusList: [PlanStatus.COMPLETE],
        }),
        currentReactivePlans: getPlansArray(state, {
          ...reactiveFilters,
          statusList: [PlanStatus.ACTIVE],
        }),
        currentRoutinePlans: getPlansArray(state, {
          ...routineFilters,
          statusList: [PlanStatus.ACTIVE],
        }),
      };
    };
  }
}
