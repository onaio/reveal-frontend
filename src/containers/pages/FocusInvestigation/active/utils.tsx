import React from 'react';
import { RouteComponentProps } from 'react-router';
import { Column } from 'react-table';
import DrillDownTableLinkedCell from '../../../../components/DrillDownTableLinkedCell';
import {
  defaultOptions,
  renderInFilterFactory,
} from '../../../../components/Table/DrillDownFilters/utils';
import { NoDataComponent } from '../../../../components/Table/NoDataComponent';
import { NO_INVESTIGATIONS_FOUND } from '../../../../configs/lang';
import { extractPlan, removeNullJurisdictionPlans } from '../../../../helpers/utils';
import { OpenSRPService } from '../../../../services/opensrp';
import { Plan } from '../../../../store/ducks/plans';

/** construct props for DrillDown for activeFI page
 * Generic type D is the type of data that will be passed to the Table
 * its not `Plan` because the plan goes through a few transformations before being rendered,
 * at render its structure will have mutated to a Dictionary courtesy of 'extractPlan'
 * @param columns - react-table compatible columns
 * @param data - array of data to be rendered in table
 * @param queryParam - the key to be used by searchForm to use as key in searchParam
 */
export const createTableProps = <D extends object, TProps extends RouteComponentProps>(
  columns: Array<Column<D>>,
  data: Plan[] | null,
  props: TProps,
  paramString: string,
  service: typeof OpenSRPService,
  noDataMessage: string = NO_INVESTIGATIONS_FOUND
) => {
  const cleanedData = data !== null ? data : [];
  const jurisdictionValidPlans = removeNullJurisdictionPlans(cleanedData);
  const thePlans = jurisdictionValidPlans.map(extractPlan);

  return {
    CellComponent: DrillDownTableLinkedCell,
    columns,
    data: thePlans,
    renderInBottomFilterBar: renderInFilterFactory({
      showColumnHider: false,
      showFilters: false,
      showPagination: true,
      showRowHeightPicker: false,
      showSearch: false,
    }),
    renderInTopFilterBar: renderInFilterFactory({
      ...defaultOptions,
      componentProps: props,
      queryParam: paramString,
      serviceClass: service,
    }),
    renderNullDataComponent: () => <NoDataComponent message={noDataMessage} />,
    useDrillDown: false,
  };
};
