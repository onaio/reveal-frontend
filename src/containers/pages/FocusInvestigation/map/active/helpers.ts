import GeojsonExtent from '@mapbox/geojson-extent';
import { Dictionary } from '@onaio/utils';
import { GREY } from '../../../../../colors';
import { GisidaProps } from '../../../../../components/GisidaWrapper';
import {
  circleLayerConfig,
  fillLayerConfig,
  lineLayerConfig,
  symbolLayerConfig,
} from '../../../../../configs/settings';
import {
  CASE_CONFIRMATION_GOAL_ID,
  GOAL_CONFIRMATION_GOAL_ID,
  LARVAL_DIPPING_ID,
  MAIN_PLAN,
  MOSQUITO_COLLECTION_ID,
  STRUCTURE_LAYER,
} from '../../../../../constants';
import { FeatureCollection } from '../../../../../helpers/utils';
import { Jurisdiction } from '../../../../../store/ducks/jurisdictions';
import { StructureGeoJSON } from '../../../../../store/ducks/structures';
import { TaskGeoJSON } from '../../../../../store/ducks/tasks';

/** FILayers - represents layers to be rendered on the FI Map */
export interface FILayers {
  features: FeatureCollection<TaskGeoJSON>;
  id: string;
  layerType: string;
  visible: boolean;
}

/** Get Gisida Wrapper Props
 * @param {Jurisdiction} jurisdiction - the jurisdiction (with geojson field)
 * @param {structures} FeatureCollection<StructureGeoJSON> - represents pre-loaded structures i.e. not added by a user
 */
export const getGisidaWrapperProps = (
  jurisdiction: Jurisdiction,
  structures: FeatureCollection<StructureGeoJSON> | null = null,
  fiLayers: FILayers[] = [],
  currentGoalId: string
): GisidaProps | null => {
  if (!jurisdiction.geojson) {
    return null;
  }
  const layers: Dictionary[] = [];

  // define line layer for Jurisdiction outline
  const jurisdictionLineLayer = {
    ...lineLayerConfig,
    id: `${MAIN_PLAN}-${jurisdiction.jurisdiction_id}`,
    source: {
      ...lineLayerConfig.source,
      data: {
        ...lineLayerConfig.source.data,
        data: JSON.stringify(jurisdiction.geojson),
      },
    },
    visible: true,
  };
  layers.push(jurisdictionLineLayer);

  // define the layer for the pre-loaded structures
  // these need both a fill and a line layer
  const structuresFillLayer = {
    ...fillLayerConfig,
    id: STRUCTURE_LAYER,
    paint: {
      ...fillLayerConfig.paint,
      'fill-color': GREY,
      'fill-outline-color': GREY,
    },
    source: {
      ...fillLayerConfig.source,
      data: {
        ...fillLayerConfig.source.data,
        data: JSON.stringify(structures),
      },
    },
    visible: true,
  };
  const structuresLineLayer = {
    ...lineLayerConfig,
    id: `${STRUCTURE_LAYER}-line`,
    paint: {
      'line-color': GREY,
      'line-opacity': 1,
      'line-width': 2,
    },
    source: {
      ...lineLayerConfig.source,
      data: {
        ...lineLayerConfig.source.data,
        data: JSON.stringify(structures),
      },
    },
  };
  layers.push(structuresFillLayer);
  layers.push(structuresLineLayer);

  let iconGoal: string = 'case-confirmation';
  switch (currentGoalId) {
    case MOSQUITO_COLLECTION_ID:
      iconGoal = 'mosquito';
      break;
    case LARVAL_DIPPING_ID:
      iconGoal = 'larval';
      break;
  }

  const goalsWithSymbols = [MOSQUITO_COLLECTION_ID, LARVAL_DIPPING_ID, CASE_CONFIRMATION_GOAL_ID];

  // deal with FI Layers
  fiLayers.forEach(element => {
    const elementIsIndexCase = element.id.includes('index-cases');
    const isGoalWithSymbol = goalsWithSymbols.includes(currentGoalId);
    if (element.layerType === 'symbol' && (isGoalWithSymbol || elementIsIndexCase)) {
      const thisLayer = {
        ...symbolLayerConfig,
        id: element.id,
        layout: {
          'icon-image': iconGoal,
          'icon-size': currentGoalId === GOAL_CONFIRMATION_GOAL_ID ? 0.045 : 0.03,
        },
        source: {
          ...symbolLayerConfig.source,
          data: {
            data: JSON.stringify(element.features),
            type: 'stringified-geojson',
          },
          type: 'geojson',
        },
        visible: element.visible,
      };
      layers.push(thisLayer);
    }

    if (element.layerType === 'circle') {
      const thisLayer = {
        ...circleLayerConfig,
        filter: ['==', '$type', 'Point'],
        id: element.id,
        paint: element.id.includes('historical-index-cases')
          ? {
              ...circleLayerConfig.paint,
              'circle-color': '#ff0000',
              // 'circle-radius': ['interpolate', ['exponential', 2], ['zoom'], 15.75, 2.5, 20.8, 50],
              'circle-stroke-color': '#ff0000',
              'circle-stroke-opacity': 1,
            }
          : {
              ...circleLayerConfig.paint,
              'circle-color': ['get', 'color'],
              'circle-stroke-color': ['get', 'color'],
              'circle-stroke-opacity': 1,
            },
        source: {
          ...circleLayerConfig.source,
          data: {
            data: JSON.stringify(element.features),
            type: 'stringified-geojson',
          },
          type: 'geojson',
        },
        visible: element.visible,
      };
      layers.push(thisLayer);
    }

    // fill layer
    if (element.layerType === 'fill') {
      const fillLayer = {
        ...fillLayerConfig,
        filter: ['==', '$type', 'Polygon'],
        id: `${element.id}-fill`,
        paint: {
          ...fillLayerConfig.paint,
          'fill-color': ['get', 'color'],
          'fill-outline-color': ['get', 'color'],
        },
        source: {
          ...fillLayerConfig.source,
          data: {
            data: JSON.stringify(element.features),
            type: 'stringified-geojson',
          },
          type: 'geojson',
        },
        visible: element.visible,
      };
      layers.push(fillLayer);
    }
    if (element.layerType === 'line') {
      const polygonLineLayer = {
        ...lineLayerConfig,
        filter: ['==', '$type', 'Polygon'],
        id: `${element.id}-fill-line`,
        paint: {
          'line-color': ['get', 'color'],
          'line-opacity': 1,
          'line-width': 2,
        },
        source: {
          ...lineLayerConfig.source,
          data: {
            data: JSON.stringify(element.features),
            type: 'stringified-geojson',
          },
          type: 'geojson',
        },
        visible: element.visible,
      };
      layers.push(polygonLineLayer);
    }
  });

  // GET BOUNDS
  // define feature collection of all geoms being rendered
  const featureCollection = { features: [jurisdiction.geojson], type: 'FeatureCollection' };
  // define bounds for gisida map position
  const bounds = GeojsonExtent(featureCollection);
  // GET BOUNDS

  const gisidaWrapperProps: GisidaProps = {
    bounds,
    currentGoal: null,
    geoData: null,
    goal: null,
    handlers: [],
    layers,
    pointFeatureCollection: null,
    polygonFeatureCollection: null,
    structures: null,
  };
  return gisidaWrapperProps;
};
