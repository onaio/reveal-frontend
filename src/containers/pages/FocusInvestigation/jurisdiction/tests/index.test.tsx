import reducerRegistry from '@onaio/redux-reducer-registry';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import flushPromises from 'flush-promises';
import { createBrowserHistory } from 'history';
import React from 'react';
import { Helmet } from 'react-helmet';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import { AN_ERROR_OCCURRED } from '../../../../../configs/lang';
import { FI_SINGLE_URL } from '../../../../../constants';
import * as errorHelpers from '../../../../../helpers/errors';
import { extractPlan } from '../../../../../helpers/utils';
import store from '../../../../../store';
import jurisdictionReducer, {
  fetchJurisdictions,
  reducerName as jurisdictionReducerName,
} from '../../../../../store/ducks/jurisdictions';
import plansReducer, {
  Plan,
  reducerName as plansReducerName,
} from '../../../../../store/ducks/plans';
import * as fixtures from '../../../../../store/ducks/tests/fixtures';
import { defaultActiveFIProps, FIJurisdiction } from '../index';

/** mock the envs */
jest.mock('../../../../../configs/env');

/**
 * This mock is necessary because FIJurisdiction renders the JurisdictionMap
 * component by passing to it a callback function that is then called with the
 * jurisdiction object.  We have to simulate this so that JurisdictionMap can have
 * a valid jurisdiction.
 */
jest.mock('../../../../JurisdictionMap', () => {
  const JurisdictionMap = (props: any) => {
    const { callback } = props;

    // Unfortunately jest does not let us access any variables defined out of this
    // scope
    callback(
      {
        geojson: {
          geometry: {
            coordinates: [
              [
                [101.166915893555, 15.0715019595332],
                [101.165628433228, 15.069429992157],
                [101.164855957031, 15.0649130333519],
                [101.164898872375, 15.061473449978],
                [101.165843009949, 15.0585311116698],
                [101.168718338013, 15.0577022766384],
                [101.173524856567, 15.0577437184666],
                [101.179447174072, 15.0583653449216],
                [101.183996200562, 15.0589455279759],
                [101.189103126526, 15.0597743581685],
                [101.191892623901, 15.0629238834779],
                [101.191549301147, 15.0671093647448],
                [101.19086265564, 15.0727036913665],
                [101.190605163574, 15.0748170653661],
                [101.188631057739, 15.0768061040682],
                [101.185412406921, 15.0769304183694],
                [101.182150840759, 15.0772619228176],
                [101.177172660828, 15.0780906816776],
                [101.174211502075, 15.0777591785211],
                [101.172151565552, 15.0765989134045],
                [101.168503761292, 15.0753557651845],
                [101.166915893555, 15.0715019595332],
              ],
            ],
            type: 'Polygon',
          },
          id: '450fc15b-5bd2-468a-927a-49cb10d3bcac',
          properties: {
            jurisdiction_name: 'TLv1_01',
            jurisdiction_parent_id: 'dad42fa6-b9b8-4658-bf25-bfa7ab5b16ae',
          },
          type: 'Feature',
        },
        jurisdiction_id: '450fc15b-5bd2-468a-927a-49cb10d3bcac',
      } // TODO: find a way to call this from fixtures
    );
    return <div>I love oov</div>;
  };
  return JurisdictionMap;
});

/** register reducers */
reducerRegistry.register(plansReducerName, plansReducer);
reducerRegistry.register(jurisdictionReducerName, jurisdictionReducer);

const history = createBrowserHistory();

describe('containers/FocusInvestigation/Jurisdiction', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders without crashing', () => {
    const mock: any = jest.fn();

    const props = {
      ...defaultActiveFIProps,
      history,
      location: mock,
      match: {
        isExact: true,
        params: { jurisdictionId: fixtures.jurisdiction1.jurisdiction_id },
        path: `${FI_SINGLE_URL}/:jurisdictionId`,
        url: `${FI_SINGLE_URL}/${fixtures.jurisdiction1.jurisdiction_id}`,
      },
    };
    shallow(
      <MemoryRouter>
        <FIJurisdiction {...props} />
      </MemoryRouter>
    );
  });

  it('renders correctly', async () => {
    store.dispatch(fetchJurisdictions([fixtures.jurisdiction1]));
    const mock: any = jest.fn();

    const supersetServiceMock: any = jest.fn();
    supersetServiceMock.mockImplementation(async () => []);

    const props = {
      ...defaultActiveFIProps,
      completeReactivePlans: [fixtures.plan101 as Plan],
      completeRoutinePlans: [fixtures.plan103 as Plan],
      currentReactivePlans: [fixtures.plan99 as Plan],
      currentRoutinePlans: [fixtures.plan102 as Plan],
      fetchPlansActionCreator: jest.fn(),
      history,
      location: mock,
      match: {
        isExact: true,
        params: { jurisdictionId: fixtures.jurisdiction1.jurisdiction_id },
        path: `${FI_SINGLE_URL}/:jurisdictionId`,
        url: `${FI_SINGLE_URL}/${fixtures.jurisdiction1.jurisdiction_id}`,
      },
      supersetService: supersetServiceMock,
    };

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <FIJurisdiction {...props} />
        </MemoryRouter>
      </Provider>
    );

    /** Show loading indicator */
    expect(wrapper.find('FIJurisdiction>Ripple').length).toEqual(1);

    await flushPromises();
    wrapper.update();

    /** No longer show loading indicator */
    expect(wrapper.find('FIJurisdiction>Ripple').length).toEqual(0);

    // check that the documents title was changed correctly
    const helmet = Helmet.peek();
    expect(helmet.title).toEqual('Focus Investigations in TLv1_01');

    expect(
      wrapper
        .find('h2.page-title')
        .last()
        .text()
    ).toEqual('Focus Investigations in TLv1_01');

    // check breadcrumbs
    expect(wrapper.find('HeaderBreadcrumb').length).toEqual(1);
    expect(wrapper.find('HeaderBreadcrumb li').length).toEqual(7);
    // check the hierarchy
    expect(toJson(wrapper.find('HeaderBreadcrumb li a'))).toMatchSnapshot('HeaderBreadcrumb li a');
    // check last item
    expect(
      wrapper
        .find('HeaderBreadcrumb li')
        .last()
        .text()
    ).toEqual('TLv1_01');

    expect(
      wrapper
        .find('.focus-area-info-title')
        .last()
        .text()
    ).toEqual('Focus Area Information');

    expect(toJson(wrapper.find('JurisdictionMap'))).toMatchSnapshot('JurisdictionMap');
    expect(toJson(wrapper.find('.focus-area-info-section dl'))).toMatchSnapshot(
      'focus-area-info-section'
    );

    expect(
      wrapper
        .find('h4.fi-table.current')
        .last()
        .text()
    ).toEqual('Current Focus Investigations');

    expect(
      wrapper
        .find('h4.fi-table.complete')
        .last()
        .text()
    ).toEqual('Complete Focus Investigations');

    // subsection titles
    expect(
      wrapper
        .find('.current-plans h3.page-title')
        .first()
        .text()
    ).toEqual('Reactive');
    expect(
      wrapper
        .find('.current-plans h3.page-title')
        .last()
        .text()
    ).toEqual('Routine');

    expect(
      wrapper
        .find('.complete-plans h3.page-title')
        .first()
        .text()
    ).toEqual('Reactive');
    expect(
      wrapper
        .find('.complete-plans h3.page-title')
        .last()
        .text()
    ).toEqual('Routine');

    expect(wrapper.find('.current-plans TableHeader').length).toEqual(2);
    expect(
      wrapper
        .find('.current-plans TableHeader')
        .first()
        .props() as any
    ).toEqual({
      plansArray: [fixtures.plan99],
    });
    expect(
      wrapper
        .find('.current-plans TableHeader')
        .last()
        .props() as any
    ).toEqual({
      plansArray: [fixtures.plan102],
    });
    expect(toJson(wrapper.find('.current-plans LinkAsButton a').first())).toMatchSnapshot(
      'Link to create new plans'
    );

    expect(
      wrapper
        .find('.current-plans DrillDownTable')
        .first()
        .props()
    ).toMatchSnapshot({
      data: expect.any(Array) /** just for purposes of making snapshot smaller */,
    });
    expect(
      wrapper
        .find('.current-plans DrillDownTable')
        .first()
        .props().data
    ).toEqual([extractPlan(fixtures.plan99 as Plan)]);

    expect(
      wrapper
        .find('.current-plans DrillDownTable')
        .last()
        .props()
    ).toMatchSnapshot({
      data: expect.any(Array) /** just for purposes of making snapshot smaller */,
    });
    expect(
      wrapper
        .find('.current-plans DrillDownTable')
        .last()
        .props().data
    ).toEqual([extractPlan(fixtures.plan102 as Plan)]);

    expect(wrapper.find('.complete-plans TableHeader').length).toEqual(2);
    expect(
      wrapper
        .find('.complete-plans TableHeader')
        .first()
        .props() as any
    ).toEqual({
      plansArray: [fixtures.plan101],
    });
    expect(
      wrapper
        .find('.complete-plans TableHeader')
        .last()
        .props() as any
    ).toEqual({
      plansArray: [fixtures.plan103],
    });
    expect(toJson(wrapper.find('.complete-plans LinkAsButton a').first())).toMatchSnapshot(
      'Link to create new plans'
    );

    expect(
      wrapper
        .find('.complete-plans DrillDownTable')
        .first()
        .props()
    ).toMatchSnapshot({
      data: expect.any(Array) /** just for purposes of making snapshot smaller */,
    });
    expect(
      wrapper
        .find('.complete-plans DrillDownTable')
        .first()
        .props().data
    ).toEqual([extractPlan(fixtures.plan101 as Plan)]);

    expect(
      wrapper
        .find('.complete-plans DrillDownTable')
        .last()
        .props()
    ).toMatchSnapshot({
      data: expect.any(Array) /** just for purposes of making snapshot smaller */,
    });
    expect(
      wrapper
        .find('.complete-plans DrillDownTable')
        .last()
        .props().data
    ).toEqual([extractPlan(fixtures.plan103 as Plan)]);
  });

  it('renders empty tables correctly', async () => {
    store.dispatch(fetchJurisdictions([fixtures.jurisdiction1]));
    const mock: any = jest.fn();

    const supersetServiceMock: any = jest.fn();
    supersetServiceMock.mockImplementation(async () => []);

    const props = {
      ...defaultActiveFIProps,
      completeReactivePlans: [],
      completeRoutinePlans: [],
      currentReactivePlans: [],
      currentRoutinePlans: [],
      fetchPlansActionCreator: jest.fn(),
      history,
      location: mock,
      match: {
        isExact: true,
        params: { jurisdictionId: fixtures.jurisdiction1.jurisdiction_id },
        path: `${FI_SINGLE_URL}/:jurisdictionId`,
        url: `${FI_SINGLE_URL}/${fixtures.jurisdiction1.jurisdiction_id}`,
      },
      supersetService: supersetServiceMock,
    };

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <FIJurisdiction {...props} />
        </MemoryRouter>
      </Provider>
    );

    await flushPromises();
    wrapper.update();

    expect(wrapper.find('.current-plans NullDataTable').length).toEqual(2);

    expect(
      wrapper
        .find('.current-plans NullDataTable h3.text-muted')
        .first()
        .text()
    ).toEqual('No Investigations Found');
    expect(
      wrapper
        .find('.current-plans NullDataTable h3.text-muted')
        .last()
        .text()
    ).toEqual('No Investigations Found');

    expect(wrapper.find('.complete-plans NullDataTable').length).toEqual(2);

    expect(
      wrapper
        .find('.complete-plans NullDataTable h3.text-muted')
        .first()
        .text()
    ).toEqual('No Investigations Found');
    expect(
      wrapper
        .find('.complete-plans NullDataTable h3.text-muted')
        .last()
        .text()
    ).toEqual('No Investigations Found');
  });

  it('renders supersetService error correctly', async () => {
    const mockDisplayError: any = jest.fn();
    (errorHelpers as any).displayError = mockDisplayError;

    const error = new Error('Catastrophy!!');

    const supersetServiceMock: any = jest.fn();
    supersetServiceMock.mockRejectedValue(error);

    const props = {
      ...defaultActiveFIProps,
      completeReactivePlans: [],
      completeRoutinePlans: [],
      currentReactivePlans: [],
      currentRoutinePlans: [],
      fetchPlansActionCreator: jest.fn(),
      history,
      location: jest.fn() as any,
      match: {
        isExact: true,
        params: { jurisdictionId: fixtures.jurisdiction1.jurisdiction_id },
        path: `${FI_SINGLE_URL}/:jurisdictionId`,
        url: `${FI_SINGLE_URL}/${fixtures.jurisdiction1.jurisdiction_id}`,
      },
      supersetService: supersetServiceMock,
    };

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <FIJurisdiction {...props} />
        </MemoryRouter>
      </Provider>
    );

    await flushPromises();
    wrapper.update();

    expect(mockDisplayError.mock.calls).toEqual([[error]]);
    expect(mockDisplayError).toHaveBeenCalledTimes(1);

    expect(toJson(wrapper.find('div'))).toMatchSnapshot('supersetService error');
  });

  it('renders fetch error correctly', async () => {
    const mockDisplayError: any = jest.fn();
    (errorHelpers as any).displayError = mockDisplayError;

    const supersetServiceMock: any = jest.fn();
    supersetServiceMock.mockImplementation(async () => undefined);

    const props = {
      ...defaultActiveFIProps,
      completeReactivePlans: [],
      completeRoutinePlans: [],
      currentReactivePlans: [],
      currentRoutinePlans: [],
      fetchPlansActionCreator: jest.fn(),
      history,
      location: jest.fn() as any,
      match: {
        isExact: true,
        params: { jurisdictionId: fixtures.jurisdiction1.jurisdiction_id },
        path: `${FI_SINGLE_URL}/:jurisdictionId`,
        url: `${FI_SINGLE_URL}/${fixtures.jurisdiction1.jurisdiction_id}`,
      },
      supersetService: supersetServiceMock,
    };

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <FIJurisdiction {...props} />
        </MemoryRouter>
      </Provider>
    );

    await flushPromises();
    wrapper.update();

    expect(mockDisplayError.mock.calls).toEqual([[new Error(AN_ERROR_OCCURRED)]]);
    expect(mockDisplayError).toHaveBeenCalledTimes(1);

    expect(toJson(wrapper.find('div'))).toMatchSnapshot('jurisdiction error');
  });
});
