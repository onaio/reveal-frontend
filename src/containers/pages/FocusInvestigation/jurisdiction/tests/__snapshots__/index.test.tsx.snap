// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/FocusInvestigation/Jurisdiction renders correctly 1`] = `
Object {
  "CellComponent": [Function],
  "columns": Array [
    Object {
      "Cell": [Function],
      "Header": "Name",
      "accessor": "plan_title",
      "minWidth": 160,
    },
    Object {
      "Header": "FI Status",
      "accessor": "plan_status",
      "id": "FI Status",
      "minWidth": 80,
    },
    Object {
      "Header": "Case Notif. Date",
      "accessor": "caseNotificationDate",
      "minWidth": 90,
    },
    Object {
      "Header": "Case Class.",
      "accessor": "caseClassification",
    },
  ],
  "data": Any<Array>,
  "hasChildren": [Function],
  "identifierField": "id",
  "linkerField": "id",
  "loading": false,
  "loadingComponent": [Function],
  "paginate": false,
  "parentIdentifierField": "parent",
  "renderNullDataComponent": [Function],
  "resize": true,
  "rootParentId": null,
  "useDrillDown": false,
}
`;

exports[`containers/FocusInvestigation/Jurisdiction renders correctly 2`] = `
Object {
  "CellComponent": [Function],
  "columns": Array [
    Object {
      "Cell": [Function],
      "Header": "Name",
      "accessor": "plan_title",
      "minWidth": 160,
    },
    Object {
      "Header": "FI Status",
      "accessor": "plan_status",
      "id": "FI Status",
      "minWidth": 80,
    },
    Object {
      "Header": "Province",
      "accessor": "province",
    },
    Object {
      "Header": "District",
      "accessor": "district",
    },
    Object {
      "Header": "Canton",
      "accessor": "canton",
    },
    Object {
      "Header": "Village",
      "accessor": "village",
    },
    Object {
      "Cell": [Function],
      "Header": "Focus Area",
      "accessor": "focusArea",
      "minWidth": 150,
    },
    Object {
      "Header": "Status",
      "accessor": "status",
      "maxWidth": 60,
    },
    Object {
      "Header": "Start Date",
      "accessor": "plan_effective_period_start",
      "minWidth": 80,
    },
    Object {
      "Header": "End Date",
      "accessor": "plan_effective_period_end",
    },
    Object {
      "Cell": [Function],
      "Header": "Action",
      "accessor": "plan_status",
      "minWidth": 70,
    },
  ],
  "data": Any<Array>,
  "hasChildren": [Function],
  "identifierField": "id",
  "linkerField": "id",
  "loading": false,
  "loadingComponent": [Function],
  "paginate": false,
  "parentIdentifierField": "parent",
  "renderNullDataComponent": [Function],
  "resize": true,
  "rootParentId": null,
  "useDrillDown": false,
}
`;

exports[`containers/FocusInvestigation/Jurisdiction renders correctly 3`] = `
Object {
  "CellComponent": [Function],
  "columns": Array [
    Object {
      "Cell": [Function],
      "Header": "Name",
      "accessor": "plan_title",
      "minWidth": 160,
    },
    Object {
      "Header": "Date Completed",
      "accessor": "date_completed",
      "minWidth": 80,
    },
    Object {
      "Header": "End Date",
      "accessor": "plan_effective_period_end",
    },
    Object {
      "Header": "Case Notif. Date",
      "accessor": "caseNotificationDate",
      "minWidth": 90,
    },
    Object {
      "Header": "Case Class.",
      "accessor": "caseClassification",
    },
  ],
  "data": Any<Array>,
  "hasChildren": [Function],
  "identifierField": "id",
  "linkerField": "id",
  "loading": false,
  "loadingComponent": [Function],
  "paginate": false,
  "parentIdentifierField": "parent",
  "renderNullDataComponent": [Function],
  "resize": true,
  "rootParentId": null,
  "useDrillDown": false,
}
`;

exports[`containers/FocusInvestigation/Jurisdiction renders correctly 4`] = `
Object {
  "CellComponent": [Function],
  "columns": Array [
    Object {
      "Cell": [Function],
      "Header": "Name",
      "accessor": "plan_title",
      "minWidth": 160,
    },
    Object {
      "Header": "Date Completed",
      "accessor": "date_completed",
      "minWidth": 80,
    },
    Object {
      "Header": "Start Date",
      "accessor": "plan_effective_period_start",
      "minWidth": 80,
    },
    Object {
      "Header": "End Date",
      "accessor": "plan_effective_period_end",
      "id": "End Date",
    },
    Object {
      "Header": "Date Completed",
      "accessor": "plan_effective_period_end",
      "id": "Date Completed",
    },
  ],
  "data": Any<Array>,
  "hasChildren": [Function],
  "identifierField": "id",
  "linkerField": "id",
  "loading": false,
  "loadingComponent": [Function],
  "paginate": false,
  "parentIdentifierField": "parent",
  "renderNullDataComponent": [Function],
  "resize": true,
  "rootParentId": null,
  "useDrillDown": false,
}
`;

exports[`containers/FocusInvestigation/Jurisdiction renders correctly: HeaderBreadcrumb li a 1`] = `
Array [
  <a
    href="/"
    onClick={[Function]}
  >
    Home
  </a>,
  <a
    href="/focus-investigation"
    onClick={[Function]}
  >
    Focus Investigations
  </a>,
  <a
    href="/focus-investigation/filter/61707fc2-c6ac-4112-a8d6-2a4861958396/236ca3fb-1b74-5028-a0c8-ab954bb28055"
    onClick={[Function]}
  >
    Trat 2
  </a>,
  <a
    href="/focus-investigation/filter/8a8e4987-95a9-4b2e-b746-3c9c3eec15c5/236ca3fb-1b74-5028-a0c8-ab954bb28055"
    onClick={[Function]}
  >
    Mueng Trat District 2
  </a>,
  <a
    href="/focus-investigation/filter/735a94f9-6ed2-4bcd-b242-8c2b5f714bee/236ca3fb-1b74-5028-a0c8-ab954bb28055"
    onClick={[Function]}
  >
    Laem Klat Canton 2
  </a>,
  <a
    href="/focus-investigation/filter/a14c6253-2d3d-4bcc-8c49-35f9e44c66e1/236ca3fb-1b74-5028-a0c8-ab954bb28055"
    onClick={[Function]}
  >
    Laem Klat Moo 8
  </a>,
]
`;

exports[`containers/FocusInvestigation/Jurisdiction renders correctly: JurisdictionMap 1`] = `
<JurisdictionMap
  callback={[Function]}
  jurisdictionId="450fc15b-5bd2-468a-927a-49cb10d3bcac"
>
  <div>
    I love oov
  </div>
</JurisdictionMap>
`;

exports[`containers/FocusInvestigation/Jurisdiction renders correctly: Link to create new plans 1`] = `
<a
  className="focus-investigation btn btn-primary float-right mt-5"
  href="/plans/new"
  onClick={[Function]}
>
  Add Focus Investigation
</a>
`;

exports[`containers/FocusInvestigation/Jurisdiction renders correctly: Link to create new plans 2`] = `
<a
  className="focus-investigation btn btn-primary float-right mt-5"
  href="/plans/new"
  onClick={[Function]}
>
  Add Focus Investigation
</a>
`;

exports[`containers/FocusInvestigation/Jurisdiction renders correctly: focus-area-info-section 1`] = `
<dl
  className="row mt-3"
>
  <dt
    className="col-4"
  >
    Province
  </dt>
  <dd
    className="col-8"
  >
    Trat 2
  </dd>
  <dt
    className="col-4"
  >
    District
  </dt>
  <dd
    className="col-8"
  >
    Mueng Trat District 2
  </dd>
  <dt
    className="col-4"
  >
    Canton
  </dt>
  <dd
    className="col-8"
  >
    Laem Klat Canton 2
  </dd>
  <dt
    className="col-4"
  >
    FI Status
  </dt>
  <dd
    className="col-8"
  >
    A1
  </dd>
  <dt
    className="col-4"
  >
    FI Reason
  </dt>
  <dd
    className="col-8"
  >
    Case Triggered
  </dd>
</dl>
`;

exports[`containers/FocusInvestigation/Jurisdiction renders fetch error correctly: jurisdiction error 1`] = `
<div>
  An error prevented this jurisdiction from loading.
</div>
`;

exports[`containers/FocusInvestigation/Jurisdiction renders supersetService error correctly: supersetService error 1`] = `
<div>
  An error prevented this jurisdiction from loading.
</div>
`;
